### PR TITLE
lsp: fix file-identity and path edge cases across the URI/vfs/loader seam

### LINKED ISSUE #248: lsp: file-identity edge cases — dot segments, path-form UNC, case-insensitive overlay bypass, BOM, id-width overflows
Follow-up from the comprehensive review of PR #239 (LSP server). Line numbers reference the PR branch at review time.

**Severity: minor (1-4), nit (5-7)** — file-identity and path edge cases across the URI/vfs/loader seam. The vfs module docs explicitly delegate canonical identity to the LSP layer ("the LSP layer derives one canonical absolute path per URI"), but several spellings of one file slip through and key separate documents.

1. **Dot segments survive decoding** (`uri.rs:62`): `file_uri_to_path` only percent-decodes and fixes drive-letter case; `/a/../b.inf` and `/b.inf` intern as two documents (stale/duplicate analyses).
2. **Path-form UNC bypasses the remote-authority rejection** (`uri.rs:55`): `file:////server/share/x.inf` parses with an empty authority and decodes to `//server/share/x.inf` — a UNC path on Windows, triggering SMB network I/O from a supposedly-local server.
3. **Case-insensitive filesystems (macOS/Windows): mis-cased import bypasses the overlay** (`loader.rs:34`): source says `use lib::Math;`, file is `lib/math.inf` — the walk derives `<root>/lib/Math.inf`, the overlay (interned under the URI spelling) misses on case-sensitive `Path` equality, but `fs::read` succeeds on the case-insensitive disk → stale disk text is analyzed, and later edits to the open buffer never invalidate the importing analysis.
4. **No UTF-8 BOM stripping in the disk ingestion path** (`loader.rs:40`): clients strip the BOM from opened buffers, but unopened closure files read via `read_to_string` keep U+FEFF — every line-0 position is off by one UTF-16 unit versus the client's view, plus a spurious lexer error at file start.
5. Scheme matching is byte-exact (`uri.rs:51`): RFC-3986-valid `File://`/`FILE://` and the RFC 8089 single-slash `file:/path` form are silently rejected.
6. Bare/malformed drive URIs (`uri.rs:119`): `file:///C:` → `C:` and `file:///c:name` → `C:name` — drive-RELATIVE paths returned instead of rejection.
7. Width overflows (robustness nits): `LineIndex::new` truncates line starts with `as u32` for ≥4 GiB texts, breaking `partition_point` monotonicity (`line_index.rs:58`); `FileId` minting wraps at 2^32 interned paths (`vfs/lib.rs:89`). Explicit failure preferable to silent wrap.

### PR DESCRIPTION
**Pull request by 0xGeorgii**
Fixes #248.

Upholds the file-identity invariant the vfs docs promise: **one on-disk file ⇒ one interned spelling**. All seven edge cases from the issue, plus comprehensive matrices.

## Changes
- **uri.rs**: dot segments removed lexically after percent-decoding (`/a/../b.inf` ≡ `/b.inf`; lexical-only, symlink caveat documented). Path-form UNC (`file:////server/share`) rejected pre-decode like a remote authority. Scheme matching is RFC-3986 case-insensitive; the RFC-8089 single-slash `file:/path` form is accepted and normalized. Bare/drive-relative drive URIs (`file:///C:`, `file:///c:name`) rejected on Windows — drive validation is anchored *before* dot-removal, because normalization strips the trailing slash that distinguishes the valid drive-root `C:/` from the invalid bare `C:`.
- **loader.rs**: on an overlay miss, the loader retries the overlay under `std::fs::canonicalize`'s on-disk spelling before falling back to disk — a mis-cased import on a case-insensitive filesystem no longer strands the analysis on stale disk text. The retry sits behind an injectable canonicalize closure so the logic is unit-tested on every platform (the closure is `std::fs::canonicalize` in production).
- **BOM**: a leading UTF-8 BOM is stripped in one shared reader (`inference::read_source_file`) used by the compiler `DiskLoader`, the IDE `VfsLoader` disk fallback, and the extern-prelude reader — BOM-prefixed files now parse instead of lexer-erroring, and closure-file positions on line 0 match the client's view.
- **Width guards**: `LineIndex::new` and `Vfs::intern` now panic explicitly with clear messages instead of silently wrapping u32 (documented via `# Panics`; not `#[should_panic]`-tested since 4 GiB texts / 2^32 paths aren't cheaply constructible).

## Testing
- 11 new uri.rs tests (dot-segment variants, UNC both forms, scheme case, single-slash, drive rejects per-platform, normalized round-trips), 6 loader tests (overlay hit/retry/fallback matrix), BOM tests at LineIndex + project level.
- Full default-workspace `cargo test`: 4378 passed, 0 failed; clippy clean on all five touched packages.

Note: touches `core/inference/src/project.rs`/`lib.rs`, so expect a small (mostly CHANGELOG) conflict against #261/#265 — whichever merges second, I'll rebase.

<h3>Confidence Score: 5/5</h3>

Safe to merge — all seven targeted edge cases have explicit tests, the new code paths are pure functions with no I/O side-effects, and the changes integrate cleanly at well-defined seams.

Every new helper in uri.rs is a pure function tested in isolation with a comprehensive matrix (11 new tests covering dot-segments, encoded UNC slashes, scheme casing, single-slash form, Windows drive rejection). The overlay-then-canonicalize retry in loader.rs uses an injectable closure so all five branches are covered by unit tests without a real filesystem. The BOM-stripping path is a single shared function exercised at the project-parse level. Width guards in LineIndex and Vfs replace silent arithmetic with explicit panics at bounds unreachable in practice.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| apps/lsp/src/uri.rs | Major URI parsing refactor: case-insensitive scheme, RFC-8089 single-slash form, dot-segment removal, post-decode UNC re-check, Windows drive validation. Logic is correct and extensively tested. |
| ide/ide-db/src/loader.rs | Adds overlay-then-canonicalize-then-disk lookup with injectable canonicalize closure; BOM stripping delegated to read_source_file. Logic and tests are correct. |
| core/inference/src/project.rs | Adds read_source_file and strip_utf8_bom as the single disk-ingestion seam; DiskLoader now routes through it. Functions are correct. |
| ide/vfs/src/lib.rs | Vfs::intern now panics explicitly instead of wrapping on u32 overflow; documented via Panics. Correct guard. |
| ide/base-db/src/line_index.rs | LineIndex::new gains an explicit assert for texts >= 4 GiB; adds a BOM-positioning test documenting expected column offsets. |
| core/inference/src/extern_prelude.rs | Single line change: read_to_string replaced by read_source_file so extern-prelude files also get BOM stripping. |
| core/inference/src/lib.rs | Re-exports read_source_file and strip_utf8_bom; minor reordering. No logic change. |
| CHANGELOG.md | Adds detailed changelog entries for all seven edge-case fixes and a reference link for issue #248. |

<h3>Sequence Diagram</h3>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant Client as LSP Client
    participant URI as uri.rs
    participant Loader as VfsLoader
    participant VFS as Vfs
    participant Disk as Disk
    Client->>URI: URI string
    URI->>URI: strip_file_scheme()
    URI->>URI: isolate_path()
    URI->>URI: early UNC reject
    URI->>URI: percent_decode()
    URI->>URI: post-decode UNC reject
    URI->>URI: normalize_path()
    URI-->>Client: canonical PathBuf or None
    Loader->>VFS: contents_of_path(path)
    alt Overlay hit
        VFS-->>Loader: "Arc<str>"
    else Miss
        Loader->>Disk: canonicalize(path)
        alt Different canonical spelling
            Loader->>VFS: contents_of_path(canonical)
            alt Hit
                VFS-->>Loader: "Arc<str>"
            else Miss
                Loader->>Disk: read_source_file
                Disk-->>Loader: BOM-stripped String
            end
        else Same spelling
            Loader->>Disk: read_source_file
            Disk-->>Loader: BOM-stripped String
        end
    end
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant Client as LSP Client
    participant URI as uri.rs
    participant Loader as VfsLoader
    participant VFS as Vfs
    participant Disk as Disk
    Client->>URI: URI string
    URI->>URI: strip_file_scheme()
    URI->>URI: isolate_path()
    URI->>URI: early UNC reject
    URI->>URI: percent_decode()
    URI->>URI: post-decode UNC reject
    URI->>URI: normalize_path()
    URI-->>Client: canonical PathBuf or None
    Loader->>VFS: contents_of_path(path)
    alt Overlay hit
        VFS-->>Loader: "Arc<str>"
    else Miss
        Loader->>Disk: canonicalize(path)
        alt Different canonical spelling
            Loader->>VFS: contents_of_path(canonical)
            alt Hit
                VFS-->>Loader: "Arc<str>"
            else Miss
                Loader->>Disk: read_source_file
                Disk-->>Loader: BOM-stripped String
            end
        else Same spelling
            Loader->>Disk: read_source_file
            Disk-->>Loader: BOM-stripped String
        end
    end
```

</a>

<sub>Reviews (3): Last reviewed commit: ["Merge branch &#39;main&#39; into 248-bug-fix-fil..."](https://github.com/inferara/inference/commit/27e69af292a436c530e41ab2c6f7ae369dfa2c6b) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=44790788)</sub>

### CODE REVIEW

File: apps/lsp/src/uri.rs [hunk 1]
```diff
@@ -42,25 +48,141 @@ pub(crate) fn from_path(path: &Path) -> Option<Uri> {
...
-    }
-    let path = &rest[path_start..];
+    let after_scheme = strip_file_scheme(uri)?;
+    let path = isolate_path(after_scheme)?;
+    if path.starts_with("//") {
+        // A `//`-prefixed path is a path-form UNC (empty authority, e.g.
+        // `file:////server/share/x`): a network path that triggers SMB I/O on
+        // Windows. Rejected like the remote-authority case.
+        return None;
+    }
```
[the remark this answers is not in the document]
Comment (0xGeorgii) [reply apps/lsp/src/uri.rs:81]: Added defense in depth rather than just documenting the gap: the decoded path is now re-checked for a `//` prefix, so `file:///%2F%2Fserver/share/x.inf` is rejected outright instead of being normalized to a local path. The pre-decode `//` check stays as a cheap early exit and its comment now notes the post-decode check closes the encoded variant, so the stated "path-form UNC is rejected" invariant holds for encoded and raw slashes alike. Covered by new tests `percent_encoded_path_form_unc_is_rejected` (both host modes plus a mixed raw/encoded spelling) and `single_encoded_slash_inside_a_path_decodes` (a lone interior `%2F` still decodes fine).
